src/Registration/cReg/ControlPointGridToDeformationConverter.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+SyneRBI Synergistic Image Reconstruction Framework (SIRF)
+Copyright 2020 University College London
+
+This is software developed for the Collaborative Computational
+Project in Synergistic Reconstruction for Biomedical Imaging (formerly CCP PETMR)
+(http://www.ccpsynerbi.ac.uk/).
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+/*!
+\file
+\ingroup Registration
+\brief Class for converting control point grids to deformation field transformations.
+
+\author Richard Brown
+\author SyneRBI
+*/
+
+#include "sirf/Reg/ControlPointGridToDeformationConverter.h"
+#include "sirf/Reg/NiftiImageData3DDeformation.h"
+#include "sirf/Reg/NiftiImageData3DBSpline.h"
+
+using namespace sirf;
+
+template<class dataType>
+ControlPointGridToDeformationConverter<dataType>::
+ControlPointGridToDeformationConverter()
+{
+    for (unsigned i=0; i<3; ++i)
+        _spacing[i] = std::numeric_limits<float>::quiet_NaN();
+}
+
+template<class dataType>
+void
+ControlPointGridToDeformationConverter<dataType>::
+set_cpg_spacing(const float spacing[3])
+{
+    for (unsigned i=0; i<3; ++i)
+        _spacing[i] = spacing[i];
+}
+
+template<class dataType>
+void
+ControlPointGridToDeformationConverter<dataType>::
+set_reference_image(const NiftiImageData<dataType> &ref)
+{
+    _template_ref_sptr = ref.clone();
+}
+
+template<class dataType>
+NiftiImageData3DDeformation<dataType>
+ControlPointGridToDeformationConverter<dataType>::
+forward(const NiftiImageData3DBSpline<dataType> &cpg)
+{
+    check_is_set_up();
+    return cpg.get_as_deformation_field(*_template_ref_sptr);
+}
+
+template<class dataType>
+NiftiImageData3DBSpline<dataType>
+ControlPointGridToDeformationConverter<dataType>::
+backward(const NiftiImageData3DDeformation<dataType> &dvf)
+{
+    check_is_set_up();
+    return NiftiImageData3DBSpline<dataType>(dvf, _spacing);
```

Comment:
it doesn't make sense. The `backward` operation has to depend on the location of the CPG, but here it only depends on the spacing. You cannot know that the CPG is in the same location as the `dvf`. I suppose it usually is of course, but would at least have to check.